### PR TITLE
Simplify webflux mono handling

### PR DESCRIPTION
**Pull request by trask**
Splitting the more problematic (and independent) part of #3798 out into a separate PR for better scrutiny/history.

Based on @HaloFour's https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3798#issuecomment-896432083, I realized that the change in #3789 that seems to fix the flakiness that we previously fixed via #3150, is not the change in instrumentation placement, but the registeration of a cancellation handler using Mono's `doOnCancel`.

### CODE REVIEW

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java [hunk 1]
```diff
@@ -35,20 +28,10 @@ public static String spanNameForHandler(Object handler) {
...
-   * Idea for this has been lifted from https://github.com/reactor/reactor-core/issues/947. Newer
-   * versions of reactor-core have easier way to access context but we want to support older
-   * versions.
-   */
-  public static <T> Function<? super Publisher<T>, ? extends Publisher<T>> finishSpanNextOrError(
-      io.opentelemetry.context.Context context) {
-    return Operators.lift(
-        (scannable, subscriber) -> new SpanFinishingSubscriber<>(subscriber, context));
+  public static <T> Mono<T> setPublisherSpan(Mono<T> mono, Context context) {
+    return mono.doOnError(t -> finishSpanIfPresent(context, t))
```
Comment (anuraaga) [inline instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java:32]: Any thoughts on `doOnError` only setting the exception and `doOnTerminate` finishing the span?

**Review by anuraaga — APPROVED** [review]

File: instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java (hunk 1 above)
Comment (trask) [reply instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java:32]: that doesn't really fit the Instrumenter model though, of calling end() and passing the exception?

Comment (trask) [reply instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java:32]: (Instrumenter coming in #3798)

Comment (anuraaga) [reply instrumentation/spring/spring-webflux-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/webflux/server/AdviceUtils.java:32]: Ah that makes sense too

**Review by laurit — APPROVED** [review]

**Pull request merged**